### DISCUSSION TITLE
Keeping a Worker alive 'forever' / managing long running processes

### DISCUSSION THREAD

**Discussion by roman-mibex-2** [com-lihaoyi/mill#3972] [Q&A] [answered]
Background: I've integrated Webpack into my build as a Mill worker. Mill does the incremental build and the the Webpack worker runs a webpack server so that things show up in the Browser.
This works well, except: Mill shut down the MillServer after some idle time and with it kills the Webpack worker with the webpack process.

So, my question is how to solve this better:

- Is there a way to keep the MillServer process 'forever' / until killed explicitly?
- Or is there a better alternative to run the webpack background process? 
  - Basic would be launch it but flag it that it is not killed?
  - There is the .runBackground facility, but that is for Java-Process only at the moment. Would that be a good fit if we adopt it to accommodate non-Java processes.
 - Maybe another obvious approach I'm missing? 

PS: Was the timeout longer in 0.11? I remember it working well in 0.11, but I'm encountering it more in 0.12. Or maybe it was an accident in 0.11, that it was kept alive by BSP requests etc.

**Comment by lihaoyi** [comment, marked as answer]
Mill currently times out after 5 minutes


https://github.com/com-lihaoyi/mill/blob/main/runner/src/mill/runner/MillServerMain.scala#L27


You should be able to increase it with a system property that you can put in `.mill-jvm-opts`. We could also increase the default; IIRC Bazel has a 30min server timeout

Making runBackground work for non-JVM processes is a TODO. We don't have that yet but I'd like to get there at some point. However, note that while `runBackground` processes will not stop when Mill shuts down, they will be restarted when Mill starts again next time and you run the command. Not sure if that behavior is acceptable for your use case?

**Reply by roman-mibex-2** [reply]
For the .runBackground: I keeping it running when Mill shuts down is acceptable.
Also running the MillServer for longer is good enough =).
